mechanics/mechanics.py

```diff
@@ -56,7 +60,16 @@ def resolve_battle(board1: Board, board2: Board) -> Tuple[int, int, List[dict]]:
         attack = randint(0, attacking_token.unit.attack)
         defense = randint(0, defending_token.unit.defense)
 
+        distance = defending_token.distance(attacking_token)
+        attack_range = attacking_token.unit.range
+        if distance <= 0:
+            raise ValueError("Distance between tokens is 0")
+
+        damage_reduction_divider = calculate_damage_divider(distance, attack_range)
+
         damage = max(0, attack - defense)
+        damage /= damage_reduction_divider
+
         defending_token.unit.hp = max(0, defending_token.unit.hp - damage)
```

Comment:
it would be nice if we could update `hp` without using `max` explicitely, e.g.:
```py
unit.hp = 100
unit.hp -= 999
assert unit.hp == 0
```
but for now it is hard to implement `hp.setter` because `Unit` is already a `@dataclass`
```